packages/base/query-field-support.ts

```diff
@@ -0,0 +1,304 @@
+import type { BaseDef, CardDef, Field } from './card-api';
+import type {
+  LooseCardResource,
+  Query,
+  Relationship,
+} from '@cardstack/runtime-common';
+import {
+  assertQuery,
+  cloneRelationship,
+  getField,
+  normalizeQueryForSignature,
+  parseQuery,
+  querySignature,
+  THIS_INTERPOLATION_PREFIX,
+  THIS_REALM_TOKEN,
+} from '@cardstack/runtime-common';
+import { initSharedState } from './shared-state';
+
+export interface QueryFieldAccessPayload {
+  instance: CardDef;
+  fieldName: string;
+  field: Field;
+}
+
+interface QueryFieldState {
+  signature?: string;
+  query?: Query;
+  searchURL?: string | null;
+  relationship?: Relationship;
+  realm?: string | null;
+  stale?: boolean;
+}
+
+export interface QueryFieldCoordinator {
+  handleQueryFieldAccess(payload: QueryFieldAccessPayload): void;
+}
+
+let queryFieldCoordinator: QueryFieldCoordinator | undefined;
+
+export function registerQueryFieldCoordinator(
+  coordinator: QueryFieldCoordinator | undefined,
+) {
+  queryFieldCoordinator = coordinator;
+}
+
+export function notifyQueryFieldAccess(instance: CardDef, field: Field) {
+  if (!field.queryDefinition || !queryFieldCoordinator) {
+    return;
+  }
+  if (isQueryFieldEvaluationInProgress(instance, field.name)) {
+    return;
+  }
```

Comment:
this is likely wrong, because we would want to cancel and restart if the inputs changed